momaland/utils/aec_wrappers.py

```diff
@@ -0,0 +1,130 @@
+
+    # https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Parallel_algorithm
+    def __init__(self, epsilon=1e-4, shape=()):
+        """Tracks the mean, variance and count of values."""
+        self.mean = np.zeros(shape, "float64")
+        self.var = np.ones(shape, "float64")
+        self.count = epsilon
+
+    def update(self, x):
+        """Updates the mean, var and count from a batch of samples."""
+        batch_mean = np.mean(x, axis=0)
+        batch_var = np.var(x, axis=0)
+        batch_count = x.shape[0]
+        self.update_from_moments(batch_mean, batch_var, batch_count)
+
+    def update_from_moments(self, batch_mean, batch_var, batch_count):
+        """Updates from batch mean, variance and count moments."""
+        self.mean, self.var, self.count = update_mean_var_count_from_moments(
+            self.mean, self.var, self.count, batch_mean, batch_var, batch_count
+        )
+
+
+def update_mean_var_count_from_moments(mean, var, count, batch_mean, batch_var, batch_count):
+    """Updates the mean, var and count using the previous mean, var, count and batch values."""
+    delta = batch_mean - mean
+    tot_count = count + batch_count
+
+    new_mean = mean + delta * batch_count / tot_count
+    m_a = var * count
+    m_b = batch_var * batch_count
+    m2 = m_a + m_b + np.square(delta) * count * batch_count / tot_count
+    new_var = m2 / tot_count
+    new_count = tot_count
+
+    return new_mean, new_var, new_count
+
+
+class NormalizeReward(Wrapper):
+    r"""This wrapper will normalize immediate rewards s.t. their exponential moving average has a fixed variance.
+
+    The exponential moving average will have variance :math:`(1 - \gamma)^2`.
+
+    Note:
+        The scaling depends on past trajectories and rewards will not be scaled correctly if the wrapper was newly
+        instantiated or the policy was changed recently.
+    """
+
+    def __init__(
+        self,
+        env,
+        agent,
+        indices,
+        gamma: float = 0.99,
+        epsilon: float = 1e-8,
+    ):
+        """This wrapper will normalize immediate rewards s.t. their exponential moving average has a fixed variance.
+
+        Args:
+            env: The environment to apply the wrapper
+            agent: the name of the agent in string whose reward(s) will be normalized.
+            indices: a ndarray with the indices for the values in the reward vector that should be normalized.
+            epsilon: A stability parameter
+            gamma: The discount factor that is used in the exponential moving average.
+        """
+        super().__init__(env)
+        self._env = env
+        self.agent = agent
+        self.indices = indices
+        self.num_rewards = self._env.reward_spaces[agent].shape[0]
+        self.return_rms = np.array(
+            [RunningMeanStd(shape=()) for _ in range(self.num_rewards)]
+        )  # separate runningmeanstd for each obj
+        self.returns = 0
```

Comment:
This seems fishy to me. Either you have only one return_rms or you have an array of returns?
The easiest implementation in my mind if to have only one index in this class and normalize only that component for the given agent.

Then if I want to normalize all rewards of an agent (we can provide helper or just let users do it in the learning script for now):
```python
for agent in env.possible_agents:
    for idx in range(env.reward_space(agent).shape[0]):
         env = NormalizeReward(env, agent, idx)
```